File: AI Challenge/code/AI/network.py
import random

class ANN:
	def __init__(self, columns:list, weights:list, autoFill=False):
		self.columns = columns
		
		# Interpret 1D weights into 3D
		self.linear = weights
		self.weights = []
		i = 0
		for x in range(1, len(columns)):
			self.weights.append( [] )
			for y1 in range(0, columns[x]):
				self.weights[x-1].append( [] )

				for y2 in range(0, columns[x-1]):
					self.weights[x-1][y1].append( weights[i] )
					i += 1

	def forward(self, data):
		if len(data) < self.columns[0]:
			raise ValueError("Invalid input data, not enough elements")

		# Fill the neurons with starting data
		neurons = []
		for x, val in enumerate(self.columns):
			neurons.append([])
			for y in range(0, val):
				if x==0:
					neurons[x].append(data[y])
				else:
					neurons[x].append(0)

		for x in range(1, len(neurons)):
			for y1 in range(0, len(neurons[x])):
				for y2 in range(0, len(neurons[x-1])):
					neurons[x][y1] = neurons[x-1][y2] * self.weights[x-1][y1][y2]
		
		# Return the last column of neurons
		i = len(neurons)-1
		return neurons[i]


	def reproduce(self, other):
		# NOTE: Variations has a small bias towards increase in neuron count

		# Determine the number of columns
		columnCount = random.choice([
			len(self.columns),
			len(other.columns)
		])
		# Mutate the result
		columnCount = round( columnCount + (random.random()*2.25-1) )
		# Ensure input and output columns exist
		if columnCount < 2:
			columnCount = 2


		# Determine the number of neurons within a column
		columns = []
		for i in range(0, columnCount):
			opts = []
			if i < len(self.columns):
				opts.append(self.columns[i])
			if i < len(other.columns):
				opts.append(other.columns[i])
			if len(opts) == 0:
				opts.append(1)

			columns.append( random.choice(opts) )
			# Mutate
			columns[i] = round( columns[i] + (random.random()*2.25-1) )
			# Ensure at least one neuron exists
			if columns[i] < 2:
				columns[i] = 2

		# Insure that input and output sizes stay the same
		columns[0] = self.columns[0]
		columns[-1] = self.columns[-1]

		
		# Determine the weights
		i = 0
		linear  = []
		weights = []
		for x in range(1, columnCount):
			weights.append([])
			for y1 in range(0, columns[x]):
				weights[x-1].append([])
				for y2 in range(0, columns[x-1]):
					opts = []

					if x < len(self.weights):
						if y1 < len(self.weights[x-1]):
							if y2 < len(self.weights[x-1][y1]):
								opts.append(self.weights[x-1][y1][y2])
					if x < len(other.weights):
						if y1 < len(other.weights[x-1]):
							if y2 < len(other.weights[x-1][y1]):
								opts.append(other.weights[x-1][y1][y2])
					if len(opts) < 1:
						opts.append(0)

					weights[x-1][y1].append( random.choice(opts) )
					weights[x-1][y1][y2] += (random.random()*2 -1)
					linear.append(weights[x-1][y1][y2])

		out = ANN([], [])
		out.columns = columns
		out.weights = weights
		out.linear  = linear
		return out

	def clone(self):
		out = ANN([], [])
		out.weights = self.weights
		out.columns = self.columns

		return out



# print( ANNForward(
# 	[3, 3, 2, 1],
# 	[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],

# 	[3, 2, 1]
# ) )

File: AI Challenge/code/AI/Cpp/network.cpp
#include "./network.hpp"


Network::Network(){
	this->topology.resize(0);
	this->weights.resize(0);
};
Network::Network(Network *other){
	this->topology = other->topology;
	this->weights = other->weights;
};
Network::Network(NetworkTopology structure){
	this->topology = structure;

	int len = structure.size()-1;
	this->weights.resize( len );

	for (int i=0; i<len; i++){
		this->weights[i].resize(structure[i+1], structure[i], 0);
	}
};
Network::Network(std::vector<Matrix> form){
	this->weights = form;

	int len = this->weights.size()+1;
	this->topology.resize(len);
	
	this->topology[0] = form[0].rows;
	for (int i=1; i<len; i++){
		this->topology[i] = form[i-1].columns;
	}
};
Network::Network(std::vector<double> form, NetworkTopology structure){
	this->topology = structure;




	// Size the weights to fit
	int len = structure.size()-1;
	this->weights.resize( len );
	for (int i=0; i<len; i++){
		this->weights[i].resize(structure[i+1], structure[i], 0);
	}


	// Transfer data
	int   ptr = 0;
	int  xLen = this->weights.size();
	int y1Len = 0;
	int y2Len = 0;
	for (int x=0; x<xLen; x++){
		y1Len = this->weights[x].rows;
		y2Len = this->weights[x].columns;
		for (int y1=0; y1<y1Len; y1++){
			for (int y2=0; y2<y2Len; y2++){
				this->weights[x][y1][y2] = form[ptr];
				ptr++;
			}
		}
	}
};




// Forward propergate the matrix through the network
Matrix Network::forward(Matrix input){
	if (input.columns != this->topology[0] || input.rows != 1){
		std::cerr << "Invalid input matrix" << std::endl;

		return Matrix(this->topology[ this->topology.size()-1 ], 1, 0);
	}

	Matrix out = input;

	int len = this->weights.size();
	int yLen = 0;
	int xLen = 0;
	for (int i=0; i<len; i++){
		out = out.dot(this->weights[i]);

		// Apply activation function
		yLen = out.rows;
		xLen = out.columns;
		for (int y=0; y<yLen; y++){
			for (int x=0; x<xLen; x++){
				out[y][x] = double(1) / (double(1) + exp(-out[y][x]));
			}
		}
	}

	return out;
};


// Create a child with this network and other
Network Network::reproduce(Network *other){
	NetworkTopology structure;
	std::vector<Matrix>  form;


	// Pick the number of columns from either parents
	// Vary the value slightly
	int      choice = random() % 2;
	int    colsSelf = this->topology.size();
	int   colsOther = other->topology.size();
	int columnCount = 0;
	if (choice == 0){
		columnCount = colsSelf;
	}else{
		columnCount = colsOther;
	}
	columnCount += int(random() % 4) - 1;
	// Ensure that input and output columns exist
	if (columnCount < 2){
		columnCount = 2;
	}
	structure.resize(columnCount);
	form.resize(columnCount-1);




	// Determine the depth for each
	std::vector<int> opts;
	int temp = 0;
	for (unsigned int x=0; x<columnCount; x++){
		opts.resize(0);

		if (x < colsSelf){
			opts.push_back(this->topology[x]);
		}
		if (x< colsOther){
			opts.push_back(other->topology[x]);
		}

		// Pick an option from either parent when valid
		if (opts.size() > 0){
			temp = opts[ random() % opts.size() ];
		}else{
			temp = 1;
		}

		// Vary
		temp += int(random() % 4) - 1;
		if (temp < 1){
			structure[x] = 1;
		}else{
			structure[x] = temp;
		}
	}
	// Ensure that the input and output sizes do not change
	structure[0]             = this->topology[0];
	structure[columnCount-1] = this->topology[colsSelf-1];




	// Determine the weights
	double delta = 0;
	for (unsigned int x=1; x<columnCount; x++){
		form[x-1].resize(structure[x], structure[x-1], 0);

		for (unsigned int y1=0; y1<structure[x-1]; y1++){
			for (unsigned int y2=0; y2<structure[x]; y2++){
				opts.resize(0);

				// Get options
				if (x < colsSelf){
					if (y1 < this->topology[x-1]){
						if (y2 < this->topology[x]){
							opts.push_back(this->weights[x-1][y1][y2]);
						}
					}
				}
				if (x < colsOther){
					if (y1 < other->topology[x-1]){
						if (y2 < other->topology[x]){
							opts.push_back(other->weights[x-1][y1][y2]);
						}
					}
				}


				// Select one
				if (opts.size() > 0){
					form[x-1][y1][y2] = opts[ random() % opts.size()];
				}

				// Vary
				delta  = random();
				delta /= RANDOM_MAX;
				form[x-1][y1][y2] += delta*0.1; // Varation rate 0.1
			}
		}
	}


	return Network(form);
};



// Helper functions for logging
int Network::length(){
	return this->topology.size();
};
int Network::neurons(){
	int tally = 0;

	int len = this->topology.size();
	for (int i=0; i<len; i++){
		tally += this->topology[i];
	}

	return tally;
};

std::string Network::toString(){
	int xLen = this->weights.size();
	int y1Len = 0;
	int y2Len = 0;
	
	int len = this->topology.size();
	std::string str = "Network ("+std::to_string(this->topology[0]);
	for (int i=1; i<len; i++){
		str += ", " + std::to_string(this->topology[i]);
	}
	str += "){\n";

	std::string temp;
	int tempLen = 0;
	for (int x=0; x<xLen; x++){
		temp = this->weights[x].toString();
		tempLen = temp.size();

		// Indent the result
		str += "  ";
		for (int i=0; i<tempLen; i++){
			if (temp[i] == '\n'){
				str += "\n  ";
			}else{
				str += temp[i];
			}
		}
		str += '\n';
	}
	str += "}";

	return str;
};



std::vector<double> Network::flatten(){
	std::vector<double> out;

	// Calculate size
	size_t size = 0;
	int len = this->topology.size();
	for (int i=1; i<len; i++){
		size += this->topology[i-1] * this->topology[i];
	}
	out.resize(size, 0);


	// Transfer data
	int   ptr = 0;
	int  xLen = this->weights.size();
	int y1Len = 0;
	int y2Len = 0;
	for (int x=0; x<xLen; x++){
		y1Len = this->weights[x].rows;
		y2Len = this->weights[x].columns;
		for (int y1=0; y1<y1Len; y1++){
			for (int y2=0; y2<y2Len; y2++){
				out[ptr] = this->weights[x][y1][y2];
				ptr++;
			}
		}
	}

	return out;
};




void Network::load(std::string filepath){
	std::vector<char> data;
	std::streampos fileSize;

	std::ifstream file(filepath, std::ifstream::binary);
	if (file.is_open()){
			file.seekg(0, std::ios_base::end);
			fileSize = file.tellg();
			file.seekg(0, std::ios_base::beg);
			data.resize(fileSize);

			// file.read(&data[0], fileSize);
			data.resize(fileSize);
			data.assign((std::istreambuf_iterator<char>(file)),
									std::istreambuf_iterator<char>());
	}else{
		std::cerr << "Failed to read network from file" << std::endl;
		return;
	}

	long long index = 0;

	std::vector<double>  form;
	NetworkTopology structure;

	std::cout << "Loaded Network ("<<data.size()<<" bytes)"<<std::endl;



	//----------------
	// Read Topology
	//----------------
	// Length of topology
	size_t topologySize = 0;
	memcpy(&topologySize, &data[index], sizeof(size_t));
	index += sizeof(size_t);
	std::cout << "  Topology ("<< topologySize <<"):";
	// Topology data
	structure.resize(topologySize);
	for (int i=0; i<topologySize; i++){
		memcpy(&structure[i], &data[index], sizeof(unsigned int));
		index += sizeof(unsigned int);
		std::cout << " " << structure[i];
	}
	std::cout << std::endl;



	//----------------
	// Read Weights
	//----------------
	// Length of weights
	size_t weightSize = 0;	
	memcpy(&weightSize, &data[index], sizeof(size_t));
	index += sizeof(size_t);
	// Weight data
	form.resize(weightSize);
	std::cout << "  Weight count: " << weightSize << std::endl;
	for (int i=0; i<weightSize; i++){
		memcpy(&form[i], &data[index], sizeof(double));
		index += sizeof(double);
	}


	*this = Network(form, structure);
};
void Network::save(std::string filepath){
	std::vector<char> data;

	auto w = this->flatten();
	// Size to fit topology
	size_t size = sizeof(size_t) + (this->topology.size()) * sizeof(unsigned int);
	// Size to fit weight data
	size += sizeof(size_t) + (w.size() * sizeof(double));
	data.resize(size);


	// Where the next chunk of data starts
	long long index = 0;


	// Write the number of elements in the topology
	size_t len = this->topology.size();
	memcpy(&data[index], &len, sizeof(size_t));
	index += sizeof(size_t);

	// Write each topology data point
	for (int i=0; i<len; i++){
		memcpy(&data[index], &this->topology[i], sizeof(unsigned int));
		index += sizeof(unsigned int);
	}


	// Write the number of elements in the topology
	len = w.size();
	memcpy(&data[index], &len, sizeof(size_t));
	index += sizeof(size_t);

	// Write each topology data point
	for (int i=0; i<len; i++){
		memcpy(&data[index], &w[i], sizeof(double));
		index += sizeof(double);
	}




	std::ofstream file(filepath, std::ios::out | std::ios::binary);
	file.write(&data[0], index);
	file.close();
	std::cout << "    Saved Network ("<< ( data.size()*sizeof(char) ) <<" bytes)" << std::endl;

	std::cout << "      Topology:";
	len = this->topology.size();
	for (int i=0; i<len; i++){
		std::cout << " " << this->topology[i];
	}
	std::cout << std::endl;
}